File: Hangman/Hangman/Gallows.cpp
#include "Gallows.h"
Gallows::Gallows(void)
	:_body_complete(false),
	_word_complete(false),
	_word(""),
	_fill_space("")
{
	//set the alphabet
	for (size_t i = 0; i < 26; ++i) {
		this->_alphabet[i] = '-';
	}

	//create empty gallows
	this->_gallows[0] = "  ======= ";
	this->_gallows[1] = "  |     |    ";
	this->_gallows[2] = "  |          ";
	this->_gallows[3] = "  |          ";
	this->_gallows[4] = "  |          ";
	this->_gallows[5] = "  |          ";
	this->_gallows[6] = "  |          ";
	this->_gallows[7] = "=====     ";
	this->_gallows[8] = "      ";
	this->_gallows[8] = this->_gallows[8] + this->_fill_space;

	//update gallows with the alphabet
	this->_gallows[1] = this->_gallows[1] + this->_alphabet[0] + ' ' + this->_alphabet[1] + ' ' + this->_alphabet[2] + ' ' + this->_alphabet[3] + ' ' + this->_alphabet[4];
	this->_gallows[2] = this->_gallows[2] + this->_alphabet[5] + ' ' + this->_alphabet[6] + ' ' + this->_alphabet[7] + ' ' + this->_alphabet[8] + ' ' + this->_alphabet[9];
	this->_gallows[3] = this->_gallows[3] + this->_alphabet[10] + ' ' + this->_alphabet[11] + ' ' + this->_alphabet[12] + ' ' + this->_alphabet[13] + ' ' + this->_alphabet[14];
	this->_gallows[4] = this->_gallows[4] + this->_alphabet[15] + ' ' + this->_alphabet[16] + ' ' + this->_alphabet[17] + ' ' + this->_alphabet[18] + ' ' + this->_alphabet[19];
	this->_gallows[5] = this->_gallows[5] + this->_alphabet[20] + ' ' + this->_alphabet[21] + ' ' + this->_alphabet[22] + ' ' + this->_alphabet[23] + ' ' + this->_alphabet[24];
	this->_gallows[6] = this->_gallows[6] + this->_alphabet[25];
}

Gallows::Gallows(std::string word)
	:_body_complete(false),
	_word_complete(false),
	_word(word),
	_fill_space(word)
{
	//create the fill space
	for (size_t i = 0; i < this->_word.length(); ++i) {
		if (this->_word[i] != ' ') {
			this->_fill_space[i] = '_';
		}
	}

	//create the alphabet
	for (size_t i = 0; i < 26; ++i) {
		this->_alphabet[i] = '-';
	}

	//create empty gallows
	this->_gallows[0] = "  ======= ";
	this->_gallows[1] = "  |     | ";
	this->_gallows[2] = "  |       ";
	this->_gallows[3] = "  |       ";
	this->_gallows[4] = "  |       ";
	this->_gallows[5] = "  |       ";
	this->_gallows[6] = "  |       ";
	this->_gallows[7] = "=====     ";
	this->_gallows[8] = "      ";
	this->_gallows[8] = this->_gallows[8] + this->_fill_space;

	//update gallows with the alphabet
	this->_gallows[1] = this->_gallows[1] + this->_alphabet[0] + ' ' + this->_alphabet[1] + ' ' + this->_alphabet[2] + ' ' + this->_alphabet[3] + ' ' + this->_alphabet[4];
	this->_gallows[2] = this->_gallows[2] + this->_alphabet[5] + ' ' + this->_alphabet[6] + ' ' + this->_alphabet[7] + ' ' + this->_alphabet[8] + ' ' + this->_alphabet[9];
	this->_gallows[3] = this->_gallows[3] + this->_alphabet[10] + ' ' + this->_alphabet[11] + ' ' + this->_alphabet[12] + ' ' + this->_alphabet[13] + ' ' + this->_alphabet[14];
	this->_gallows[4] = this->_gallows[4] + this->_alphabet[15] + ' ' + this->_alphabet[16] + ' ' + this->_alphabet[17] + ' ' + this->_alphabet[18] + ' ' + this->_alphabet[19];
	this->_gallows[5] = this->_gallows[5] + this->_alphabet[20] + ' ' + this->_alphabet[21] + ' ' + this->_alphabet[22] + ' ' + this->_alphabet[23] + ' ' + this->_alphabet[24];
	this->_gallows[6] = this->_gallows[6] + this->_alphabet[25];
}

Gallows::Gallows(const Gallows& gallows)
	:_body_complete(gallows._body_complete),
	_word_complete(gallows._word_complete),
	_word(gallows._word),
	_fill_space(gallows._fill_space)
{
	for (int i = 0; i < GALLOWS_SIZE; ++i) {
		this->_gallows[i] = gallows._gallows[i];
	}

	for (int i = 0; i < ALPHABT_SIZE; ++i) {
		this->_alphabet[i] = gallows._alphabet[i];
	}
}

Gallows::~Gallows(void)
{  }

Gallows& Gallows::operator = (const Gallows& gallows) {
	this->_body_complete = gallows._body_complete;
	this->_word_complete = gallows._word_complete;
	this->_word = gallows._word;
	this->_fill_space = gallows._fill_space;

	for (int i = 0; i < GALLOWS_SIZE; ++i) {
		this->_gallows[i] = gallows._gallows[i];
	}

	for (int i = 0; i < ALPHABT_SIZE; ++i) {
		this->_alphabet[i] = gallows._alphabet[i];
	}
}

void Gallows::set_word(std::string word) {
	//word can only be set if it wasn't set previously
	//(i.e. set word when fill space is not set)
	this->_word = word;
	this->_fill_space = word;

	for (size_t i = 0; i < this->_word.length(); ++i) {
		if (this->_word[i] != ' ') {
			this->_fill_space[i] = '_';
		}
	}

	//make sure the fill space is seen on the board
	this->_gallows[8] = this->_gallows[8] + this->_fill_space;
}

std::string* Gallows::get_gallows(void) {
	return this->_gallows;
}

bool Gallows::is_body_complete(void) {
	return this->_body_complete;
}

bool Gallows::is_word_complete(void) {
	return this->_word_complete;
}

void Gallows::update_row1(void) {
	this->_gallows[1] = this->_gallows[1] + this->_alphabet[0] + ' ' + this->_alphabet[1] + ' ' + this->_alphabet[2] + ' ' + this->_alphabet[3] + ' ' + this->_alphabet[4];
}

void Gallows::create_head(void) {
	this->_gallows[2] = "  |     O ";
	this->_gallows[2] = this->_gallows[2] + this->_alphabet[5] + ' ' + this->_alphabet[6] + ' ' + this->_alphabet[7] + ' ' + this->_alphabet[8] + ' ' + this->_alphabet[9];
}

void Gallows::create_body(void) {
	this->_gallows[3] = "  |     | ";
	this->_gallows[4] = "  |     | ";

	this->_gallows[3] = this->_gallows[3] + this->_alphabet[10] + ' ' + this->_alphabet[11] + ' ' + this->_alphabet[12] + ' ' + this->_alphabet[13] + ' ' + this->_alphabet[14];
	this->_gallows[4] = this->_gallows[4] + this->_alphabet[15] + ' ' + this->_alphabet[16] + ' ' + this->_alphabet[17] + ' ' + this->_alphabet[18] + ' ' + this->_alphabet[19];
}

void Gallows::create_left_arm(void) {
	//the left arm can only be created when the body already exists
	this->_gallows[3] = "  |    /| ";
	this->_gallows[3] = this->_gallows[3] + this->_alphabet[10] + ' ' + this->_alphabet[11] + ' ' + this->_alphabet[12] + ' ' + this->_alphabet[13] + ' ' + this->_alphabet[14];
}

void Gallows::create_right_arm(void) {
	//the right can can only be created when the left arm exists
	this->_gallows[3] = "  |    /|\\";
	this->_gallows[3] = this->_gallows[3] + this->_alphabet[10] + ' ' + this->_alphabet[11] + ' ' + this->_alphabet[12] + ' ' + this->_alphabet[13] + ' ' + this->_alphabet[14];
}

void Gallows::create_left_leg(void) {
	this->_gallows[5] = "  |    /  ";
	this->_gallows[5] = this->_gallows[5] + this->_alphabet[20] + ' ' + this->_alphabet[21] + ' ' + this->_alphabet[22] + ' ' + this->_alphabet[23] + ' ' + this->_alphabet[24];
}

void Gallows::create_right_leg(void) {
	//the right leg can only be created when the left leg exists
	this->_gallows[5] = "  |    / \\";
	this->_gallows[5] = this->_gallows[5] + this->_alphabet[20] + ' ' + this->_alphabet[21] + ' ' + this->_alphabet[22] + ' ' + this->_alphabet[23] + ' ' + this->_alphabet[24];

	//the body is complete when the right leg is created
	this->_body_complete = true;
}

void Gallows::update_row6(void) {
	this->_gallows[6] = this->_gallows[6] + this->_alphabet[25];
}

File: hangman/Gallows.py
global BOARD_SIZE
global GALLOWS_SIZE
global ALPHABT_SIZE

BOARD_SIZE = 10
GALLOWS_SIZE = 8
ALPHABT_SIZE = 26

class Gallows:

    def __init__(self):
        self.__body_complete = False
        self.__word_complete = False
        self.__board = [''] * BOARD_SIZE
        self.__gallows = [''] * GALLOWS_SIZE
        self.__alphabet = ['-'] * ALPHABT_SIZE
        self.__word = ''
        self.__fill_space = ''
        self.__attempt = 0

        #create empty gallows
        self.__gallows[0] = '  ======= '
        self.__gallows[1] = '  |     |    '
        self.__gallows[2] = '  |          '
        self.__gallows[3] = '  |          '
        self.__gallows[4] = '  |          '
        self.__gallows[5] = '  |          '
        self.__gallows[6] = '  |          '
        self.__gallows[7] = '=====     '

        #update the whole board with the gallows and the alphabet
        self.__update_board()
        self.__update_fill_space()
        self.__board[9] = ' '
    
    def user_guess(self, c):
        index, last_pos = 0, 0
        found_letter = False

        #if the letter is found, replace every occurance
        #of the letter in fill space
        while index > -1:
            index = self.__word.find(c, last_pos)

            if index > -1:
                self.__board[9] = 'Letter found! :)'
                self.__fill_space = self.__fill_space[:index] + c + self.__fill_space[(index+1):]
                last_pos = index + 1
                found_letter = True
            elif not found_letter:
                self.__board[9] = 'Letter not found :('
                self.__attempt += 1

        #mark the letter in the alphabet list
        self.__alphabet[ord(c) - 65] = c

        #mark the proper body part if the letter was not found
        #(i.e. found_letter is true)
        if not found_letter:
            if self.__attempt == 1:
                self.__create_head()

            elif self.__attempt == 2:
                self.__create_body()
            
            elif self.__attempt == 3:
                self.__create_left_arm()

            elif self.__attempt == 4:
                self.__create_right_arm()

            elif self.__attempt == 5:
                self.__create_left_leg()

            elif self.__attempt == 6:
                self.__create_right_leg()

        #check if the word was completed
        self.__word_complete = self.__fill_space.find('_') < 0

        #update the whole board
        self.__update_board()

        #update the fill_space
        self.__update_fill_space()

    def user_guess_line(self, line):
        if line == self.__word:
            self.__fill_space = self.__word
        else:
            self.__attempt = 6
            self.__create_head()
            self.__create_body()
            self.__create_left_arm()
            self.__create_right_arm()
            self.__create_left_leg()
            self.__create_right_leg()

        #check if the word was completed
        self.__word_complete = self.__fill_space.find('_') < 0

        #update the whole board
        self.__update_board()

        #update the fill_space
        self.__update_fill_space()

    def set_word(self, word):
        #word can only be set if it wasn't set previously
        #(i.e. set word when fill space is not set)
        self.__word = word
        self.__fill_space = ''

        for i in range(len(self.__word)):
            if self.__word[i] != ' ':
                self.__fill_space += '_'
            else:
                self.__fill_space += self.__word[i]

        #make sure the fill space is seen on the board
        self.__update_fill_space()

    def get_board(self):
        return self.__board

    def is_body_complete(self):
        return self.__body_complete

    def is_word_complete(self):
        return self.__word_complete
	
    #the below methods are supposed to be 'private' and thus
    #the leading underscores
    def __update_board(self):
        self.__board[0] = self.__gallows[0]
        self.__board[1] = self.__gallows[1] + self.__alphabet[0] + ' ' + self.__alphabet[1] + ' ' + self.__alphabet[2] + ' ' + self.__alphabet[3] + ' ' + self.__alphabet[4]
        self.__board[2] = self.__gallows[2] + self.__alphabet[5] + ' ' + self.__alphabet[6] + ' ' + self.__alphabet[7] + ' ' + self.__alphabet[8] + ' ' + self.__alphabet[9]
        self.__board[3] = self.__gallows[3] + self.__alphabet[10] + ' ' + self.__alphabet[11] + ' ' + self.__alphabet[12] + ' ' + self.__alphabet[13] + ' ' + self.__alphabet[14]
        self.__board[4] = self.__gallows[4] + self.__alphabet[15] + ' ' + self.__alphabet[16] + ' ' + self.__alphabet[17] + ' ' + self.__alphabet[18] + ' ' + self.__alphabet[19]
        self.__board[5] = self.__gallows[5] + self.__alphabet[20] + ' ' + self.__alphabet[21] + ' ' + self.__alphabet[22] + ' ' + self.__alphabet[23] + ' ' + self.__alphabet[24]
        self.__board[6] = self.__gallows[6] + self.__alphabet[25]
        self.__board[7] = self.__gallows[7]

    def __update_fill_space(self):
        self.__board[8] = '      '
        self.__board[8] = self.__board[8] + self.__fill_space

    def __create_head(self):
        self.__gallows[2] = '  |     O    '

    def __create_body(self):
        self.__gallows[3] = '  |     |    '
        self.__gallows[4] = '  |     |    '

    def __create_left_arm(self):
        #the left arm can only be created when the body already exists
        self.__gallows[3] = '  |    /|   '

    def __create_right_arm(self):
        #the right can can only be created when the left arm exists
        self.__gallows[3] = '  |    /|\\   '

    def __create_left_leg(self):
        self.__gallows[5] = '  |    /     '

    def __create_right_leg(self):
        #the right leg can only be created when the left leg exists
        self.__gallows[5] = '  |    / \\   '

        #the body is complete when the right leg is created
        self.__body_complete = True

